### PR TITLE
New logtest endpoints and API integration tests

### LINKED ISSUE #5577: [API] Add new endpoints to support logtest.
|Working branch| 
|---|
|[feature/5577-new-logtest-endpoints](https://github.com/wazuh/wazuh/tree/feature/5577-new-logtest-endpoints)|

|Wazuh version|Component|Install type|Install method|
|---|---|---|---|
| 4.1.0 | API | Manager | Packages/Sources |

# Description
Hello team. As explained in https://github.com/wazuh/wazuh/issues/5337, the logtest tool will undergo multiple changes to allow being used in the API and the in the APP. The specifications of this issue may still change depending on future decisions / revisions.
​
New endpoints are required:
- `PUT /logtest`: Check if a specified log raises any alert among other information.
- `DELETE /logtest/sessions/{token}`: Delete the saved session corresponding to {token}
​

They will be **distributed API call**, the worker nodes do not test events, the test is sent to the master and returns results.
​

`PUT /logtest` should accept the following list of parameters as a RequestBody:
- `token`:  alphanumeric string.
- `log_format`:  Enum of strings. It must accept any of these values https://documentation.wazuh.com/3.13/user-manual/reference/ossec-conf/localfile.html#log-format 
- `location`: path string (_paths).
- `event`: string
​

`DELETE /logtest/sessions/{token}` should accept the following parameter, in path:
- `token`: alphanumeric string
​

​
The output will be in JSON format like the one you can see here, and will be returned as is by the socket. It is probably still undergoing changes as it has to return a token that does not appear right now:
```JSON
{
    "token": "bd83b45f",
    "messages": ["INFO: (7202): Session initialized with token 'bd83b45f'"],
    "output": {
        "timestamp": "2020-08-27T08:36:54.740+0000",
        "rule": {
            "description": "Systemd rules",
            "id": "40700",
            "firedtimes": 1,
            "mail": false,
            "groups": ["local", "systemd"]
        },
        "agent": {
            "id": "000",
            "name": "wazuh-master"
        },
        "manager": {
            "name": "wazuh-master"
        },
        "id": "1598517414.1892559",
        "cluster": {
            "name": "wazuh",
            "node": "master-node"
        },
        "full_log": "Jun 24 11:54:19 Master systemd[2099]: Started VTE child process 20118 launched by terminator process 17756.",
        "predecoder": {
            "program_name": "systemd",
            "timestamp": "Jun 24 11:54:19",
            "hostname": "Master"
        },
        "decoder": {
            "name": "systemd"
        },
        "location": "master->/var/log/syslog"
    },
    "alert": false,
    "codemsg": 0
}
```

​
Kind regards,
Selu.

### PR DESCRIPTION
**Pull request by mcarmona99**
|Related issue|
|---|
|#5577|

Hi team,

This PR closes #5577 . I have added the new logtest endpoints `PUT /logtest` and `DELETE /logtest/sessions/{token}`. 

Summary of the changes and new files:

- New: `logtest_controller.py`.
- Changes on `changes.md` in order to include the 2 new endpoints.
- Now, `spec.yaml` includes new entries in the `x-rbac-catalog`, the request body for `PUT /logtest` endpoint, the new endpoints and their parameters.
- `framework/wazuh/tests/data/security/rbac_catalog.yml` includes the new rbac options for the new endpoints, test passed:
```
pytest framework/wazuh/tests/test_security.py 
============= 65 passed, 12 warnings in 69.19s (0:01:09) =============
```
- As mentioned above, `PUT /logtest` uses a RequestBody, new logtest model added to `api/api/models/logtest_model.py` in order to check it.
- New API integration test cases. Created `api/test/integration/test_logtest_endpoints.tavern.yaml` and tests cases to check that all functionalities of the `logtest` tool work properly.

### CODE REVIEW

**Comment by mcarmona99** [comment]
In the last commits, I have updated the logtest API integration test and added the rbac logtest tests.

TEST RESULTS:

test_logtest_endpoints.tavern.yaml 
	 2 passed, 4 warnings

test_rbac_black_logtest_endpoints.tavern.yaml 
	 2 passed, 4 warnings

test_rbac_white_logtest_endpoints.tavern.yaml 
	 1 passed, 3 warnings

File: api/test/integration/test_logtest_endpoints.tavern.yaml [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        log_format: "syslog"
+        location: "master->/var/log/syslog"
+    response:
+      status_code: 200
+      json:
+        data:
+          token: !anystr
+          output:
+            full_log: "Jun 24 11:54:19 Master systemd[2099]: Started VTE child process 20118 launched by terminator process 17756."
+            location: "master->/var/log/syslog"
```
Comment (vicferpoy) [inline api/test/integration/test_logtest_endpoints.tavern.yaml:45]: Is this necessary? Consider adding all the expected keys and using `!anystr`

File: api/test/integration/test_logtest_endpoints.tavern.yaml [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+          codemsg: 1 # codemsg = 1 because there is a warning (token generated)
```
Comment (vicferpoy) [inline api/test/integration/test_logtest_endpoints.tavern.yaml:46]: Same as above with `!anyint`.

File: api/test/integration/test_logtest_endpoints.tavern.yaml [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+      save:
+        json:
+          returned_token_2: data.token
```
Comment (vicferpoy) [inline api/test/integration/test_logtest_endpoints.tavern.yaml:125]: Change the names of the tokens to make them more descriptive. The first one could be `returned_token` and this one could be `delete_session_token` for instance.

File: api/test/integration/test_logtest_endpoints.tavern.yaml [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        location: "master->/var/log/syslog"
+    response:
+      status_code: 200
+      json:
+        data:
+          token: !anystr
+          output:
+            full_log: "Jun 24 11:54:19 Master systemd[2099]: Started VTE child process 20118 launched by terminator process 17756."
+            location: "master->/var/log/syslog"
+          codemsg: 1 # codemsg = 1 because there is a warning (token generated)
```
Comment (vicferpoy) [inline api/test/integration/test_logtest_endpoints.tavern.yaml:122]: This is not necessary as this stage purpose is only to get a token.

File: api/test/integration/test_logtest_endpoints.tavern.yaml [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+      verify: False
+      url: "{protocol:s}://{host:s}:{port:d}/logtest/sessions/{returned_token_2:s}"
+      method: DELETE
+      headers:
+        Authorization: "Bearer {test_login_token}"
+    response:
+      status_code: 200
+      json:
+        data:
+          codemsg: -1
```
Comment (vicferpoy) [inline api/test/integration/test_logtest_endpoints.tavern.yaml:164]: Add a comment to explain why this code is `-1`.

File: api/test/integration/test_rbac_black_logtest_endpoints.tavern.yaml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+  - name: End logtest session (deny)
+    request:
+      verify: False
+      url: "{protocol:s}://{host:s}:{port:d}/logtest/sessions/token"
+      method: DELETE
+      headers:
+        Authorization: "Bearer {test_login_token}"
+    response:
+      <<: *permission_denied
```
Comment (vicferpoy) [inline api/test/integration/test_rbac_black_logtest_endpoints.tavern.yaml:42]:
```suggestion
      <<: *permission_denied
    
```

File: api/test/integration/test_rbac_white_logtest_endpoints.tavern.yaml [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      verify: False
+      url: "{protocol:s}://{host:s}:{port:d}/logtest/sessions/{returned_token:s}"
+      method: DELETE
+      headers:
+        Authorization: "Bearer {test_login_token}"
+    response:
+      status_code: 200
+      json:
+        data:
+          codemsg: 0
```
Comment (vicferpoy) [inline api/test/integration/test_rbac_white_logtest_endpoints.tavern.yaml:41]:
```suggestion
          codemsg: 0
        
```

**Review by vicferpoy — CHANGES_REQUESTED** [review]
Update the changelog for `4.1`

File: api/api/spec/spec.yaml [hunk 8]
```diff
@@ -8871,6 +8918,114 @@ paths:
...
+      operationId: api.controllers.logtest_controller.end_logtest_session
+      x-rbac-actions:
+        - $ref: '#/x-rbac-catalog/actions/logtest:end_session'
+      parameters:
+        - $ref: '#/components/parameters/logtest_token'
+        - $ref: '#/components/parameters/pretty'
+        - $ref: '#/components/parameters/wait_for_complete'
+      responses:
+        '200':
+          description: "Successfully deleted the saved session."
```
[the remark this answers is not in the document]
Comment (mcarmona99) [reply api/api/spec/spec.yaml]: Done

**Comment by mcarmona99** [comment]
**Test results after the review changes:**

```
test_logtest_endpoints.tavern.yaml 
	 2 passed, 4 warnings

test_rbac_black_logtest_endpoints.tavern.yaml 
	 2 passed, 4 warnings

test_rbac_white_logtest_endpoints.tavern.yaml 
	 1 passed, 3 warnings
```

**Comment by mcarmona99** [comment]
**Test results after framework review changes**

In order to make the logtest rbac white test pass, I have used the default rbac configuration. Something similar was implemented for the mitre rbac test.

- API integration tests

```
test_logtest_endpoints.tavern.yaml 
	 2 passed, 4 warnings
	 
test_rbac_black_logtest_endpoints.tavern.yaml 
	 2 passed, 4 warnings
	 
test_rbac_white_logtest_endpoints.tavern.yaml 
	 1 passed, 3 warnings
```


- Unittests

```
framework/wazuh/tests/test_logtest.py ......                                                      [100%]
=========================================== 6 passed in 0.75s ===========================================

framework/wazuh/core/tests/test_logtest.py ..                                                     [100%]
=========================================== 2 passed in 0.06s ===========================================
```

File: api/test/integration/test_logtest_endpoints.tavern.yaml [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+          token: "{returned_token}"
+          codemsg: 0
+
+
+---
+test_name: DELETE /logtest/sessions
+
+stages:
+
+  - name: Run logtest in order to generate token
```
Comment (davidjiglesias) [inline api/test/integration/test_logtest_endpoints.tavern.yaml:93]:
```suggestion
  - name: Run logtest in order to generate the session token
```

File: api/test/integration/test_logtest_endpoints.tavern.yaml [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    response:
+      status_code: 200
+      json:
+        data:
+          token: !anystr
+      save:
+        json:
+          delete_session_token: data.token
+
+  - name: Try to end logtest session of an invalid token
```
Comment (davidjiglesias) [inline api/test/integration/test_logtest_endpoints.tavern.yaml:114]:
```suggestion
  - name: Try to end logtest session using an invalid token
```

File: api/test/integration/test_logtest_endpoints.tavern.yaml [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+      method: PUT
+      headers:
+        Authorization: "Bearer {test_login_token}"
+        content-type: application/json
+      json:
+        invalid_key: "token"
+    response:
+      status_code: 400
+
+  - name: Run logtest with an invalid token
```
Comment (davidjiglesias) [inline api/test/integration/test_logtest_endpoints.tavern.yaml:23]: Why this returns a 200 response with all ok if it is using an invalid token?

File: api/test/integration/test_rbac_black_logtest_endpoints.tavern.yaml [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        location: "master->/var/log/syslog"
+    response: &permission_denied
+      status_code: 403
+      json:
+        error: 4000
+
+---
+test_name: DELETE /logtest/sessions
+
+marks:
```
Comment (davidjiglesias) [inline api/test/integration/test_rbac_black_logtest_endpoints.tavern.yaml:29]: I am almost sure we dont use marks in every test but just once for the entire file, confirm please

File: api/test/integration/test_rbac_white_logtest_endpoints.tavern.yaml [hunk 13]
```diff
@@ -0,0 +1,41 @@
+---
+test_name: PUT /logtest and DELETE /logtest/sessions/{token}
```
Comment (davidjiglesias) [inline api/test/integration/test_rbac_white_logtest_endpoints.tavern.yaml:2]: Separate tests for different endpoints (useful for documentation and reporting)

File: api/api/spec/spec.yaml [hunk 14]
```diff
@@ -8987,6 +9026,118 @@ paths:
...
+        '413':
+          $ref: '#/components/responses/RequestTooLargeResponse'
+        '429':
+          $ref: '#/components/responses/TooManyRequestsResponse'
+
+  /logtest/sessions/{token}:
+    delete:
+      tags:
+        - Logtest
+      summary: "Delete a saved session"
```
Comment (davidjiglesias) [inline api/api/spec/spec.yaml:9104]:
```suggestion
      summary: "End session"
```

File: api/api/spec/spec.yaml [hunk 15]
```diff
@@ -8987,6 +9026,118 @@ paths:
...
+      description: "Delete the saved session corresponding to {token}"
```
Comment (davidjiglesias) [inline api/api/spec/spec.yaml:9105]:
```suggestion
      description: "Delete the saved logtest session corresponding to {token}"
```

**Review by davidjiglesias — CHANGES_REQUESTED** [review]

**Comment by mcarmona99** [comment]
Test results after the review changes:

```
test_logtest_endpoints.tavern.yaml 
	 2 passed, 4 warnings

test_rbac_black_logtest_endpoints.tavern.yaml 
	 2 passed, 4 warnings

test_rbac_white_logtest_endpoints.tavern.yaml 
	 1 passed, 3 warnings
```

File: api/api/controllers/logtest_controller.py [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+    -------
+    Data
+        Logtest result after analyzing the event.
+    """
+    Body.validate_content_type(request, expected_content_type='application/json')
+    f_kwargs = await LogtestModel.get_kwargs(request)
+
+    dapi = DistributedAPI(f=logtest.run_logtest,
+                          f_kwargs=remove_nones_to_dict(f_kwargs),
+                          request_type='local_master',
```
[the remark this answers is not in the document]
Comment (mcarmona99) [reply api/api/controllers/logtest_controller.py:38]: When the `logtest` tool was designed, the core team decided that the request type would be `local_master`.

File: api/test/integration/test_logtest_endpoints.tavern.yaml (hunk 11 above)
Comment (mcarmona99) [reply api/test/integration/test_logtest_endpoints.tavern.yaml]: It is using a token with no logtest session. In the response, the new token for the new session is shown.
I have changed the test name in order to make it easier to understand.

File: api/test/integration/test_rbac_black_logtest_endpoints.tavern.yaml (hunk 12 above)
Comment (mcarmona99) [reply api/test/integration/test_rbac_black_logtest_endpoints.tavern.yaml]: Confirmed, I made a mistake. Sorry for the confusion.

File: api/test/integration/test_rbac_white_logtest_endpoints.tavern.yaml (hunk 13 above)
Comment (mcarmona99) [reply api/test/integration/test_rbac_white_logtest_endpoints.tavern.yaml]: Tests separated and new `PUT /logtest` test case added.

**Review by davidjiglesias — APPROVED** [review]

**Pull request merged**